src/db/delete_crate.rs

```diff
@@ -0,0 +1,179 @@
+use super::file::{s3_client, S3_BUCKET_NAME};
+use failure::Error;
+use postgres::Connection;
+use rusoto_s3::{DeleteObjectsRequest, ListObjectsV2Request, ObjectIdentifier, S3Client, S3};
+
+/// List of directories in docs.rs's underlying storage (either the database or S3) containing a
+/// subdirectory named after the crate. Those subdirectories will be deleted.
+static STORAGE_PATHS_TO_DELETE: &[&str] = &["rustdoc", "sources"];
+
+#[derive(Debug, Fail)]
+enum CrateDeletionError {
+    #[fail(display = "crate is missing: {}", _0)]
+    MissingCrate(String),
+}
+
+pub fn delete_crate(conn: &Connection, name: &str) -> Result<(), Error> {
+    let crate_id_res = conn.query("SELECT id FROM crates WHERE name = $1", &[&name])?;
+    let crate_id = if crate_id_res.is_empty() {
+        return Err(CrateDeletionError::MissingCrate(name.into()).into());
+    } else {
+        crate_id_res.get(0).get("id")
+    };
+
+    delete_from_database(conn, name, crate_id)?;
+    if let Some(s3) = s3_client() {
+        delete_from_s3(&s3, name)?;
+    }
+
+    Ok(())
+}
+
+fn delete_from_database(conn: &Connection, name: &str, crate_id: i32) -> Result<(), Error> {
+    let transaction = conn.transaction()?;
+
+    transaction.execute(
+        "DELETE FROM sandbox_overrides WHERE crate_name = $1",
+        &[&name],
+    )?;
+    transaction.execute(
```

Comment:
I was looking for a way to use [`batch_execute`](https://docs.rs/postgres/0.15.2/postgres/transaction/struct.Transaction.html#method.batch_execute) with parameterized queries but it doesn't look like it's possible. Maybe we could query the crate id once and use `batch_execute` with a `format!` string? That doesn't seem too outrageous ... if you don't think it's worth the effort it's not a big deal though.

Reply:
I'd prefer not to use `format!()`s just in case. Also optimizing 6 queries into one for a CLI program called every once in a few months doesn't seem too big of a priority.